cmds/networkd/addr_watcher.go

```diff
@@ -0,0 +1,150 @@
+package main
+
+import (
+	"context"
+
+	"github.com/rs/zerolog/log"
+
+	"time"
+
+	"github.com/cenkalti/backoff/v3"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/network"
+	"github.com/threefoldtech/zos/pkg/network/ifaceutil"
+	"github.com/threefoldtech/zos/pkg/network/types"
+	"github.com/vishvananda/netlink"
+)
+
+type WatchedLinks struct {
+	linkNames map[string]struct{}
+	bcdb      network.TNoDB
+	nodeID    pkg.Identifier
+}
+
+func NewWatchedLinks(linkNames []string, nodeID pkg.Identifier, db network.TNoDB) WatchedLinks {
+	names := make(map[string]struct{}, len(linkNames))
+
+	for _, n := range linkNames {
+		names[n] = struct{}{}
+	}
+
+	return WatchedLinks{
+		linkNames: names,
+		bcdb:      db,
+		nodeID:    nodeID,
+	}
+}
+
+func (w WatchedLinks) CallBack(update netlink.AddrUpdate) error {
+	link, err := netlink.LinkByIndex(update.LinkIndex)
+	if err != nil {
+		return nil
+	}
+
+	// skip link that are not watched
+	if _, ok := w.linkNames[link.Attrs().Name]; !ok {
+		return nil
+	}
+
+	log.Info().Msg("send network interfaces update to BCDB")
+
+	ifaces, err := getLocalInterfaces()
+	if err != nil {
+		return err
+	}
+
+	return publishIfaces(ifaces, w.nodeID, w.bcdb)
+}
+
+func getLocalInterfaces() ([]types.IfaceInfo, error) {
+	var output []types.IfaceInfo
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		log.Error().Err(err).Msgf("failed to list interfaces")
+		return nil, err
+	}
+
+	for _, link := range ifaceutil.LinkFilter(links, []string{"device", "bridge"}) {
+		//TODO: why bringing the interface up? shouldn't we just check if it's up or not?
+		if err := netlink.LinkSetUp(link); err != nil {
```

Comment:
but not sure if it waits for the syscall to complete (i.e. MII detected), or it 'just' sends the syscall, and continues it's merry way. If the latter, we should wait timeout time to see if there is effectively a link.